### PR TITLE
T-12 Add portfolio for each new user

### PR DESCRIPTION
**Pull request by AntonBessarabov**
When user is created. portofolio entity should be generated for him/her

Configured owin identity with Autofac.
Added Create() method to PortfolioService.
Added portfolio creation when user created.
Added tests for PortfolioService.

### CODE REVIEW

**Comment by AntonBessarabov** [comment]
[T-12 Add portfolio for each new user](https://trello.com/c/QPPqJVx3/12-t-12-add-portfolio-for-each-new-user)

File: PortfolioKeeper/App_Start/AutofacConfig.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                .As<IUnitOfWork>()
+                .InstancePerRequest();
+
+            builder.RegisterType<RepositoryFactory>()
+                .As<IRepositoryFactory>()
+                .InstancePerRequest();
+
+            #region repositories
+
+            builder.RegisterType<BaseRepository<Asset, int>>()
```
Comment (goremukin) [inline PortfolioKeeper/App_Start/AutofacConfig.cs]: why not?
```
builder.RegisterGeneric(typeof(BaseRepository<,>))
                .As(typeof(IRepository<,>))
                .InstancePerRequest();
```

File: PortfolioKeeper/App_Start/Startup.Auth.cs [hunk 2]
```diff
@@ -14,10 +14,6 @@ public partial class Startup
     {
         public void ConfigureAuth(IAppBuilder app)
         {
-            app.CreatePerOwinContext(PortfolioKeeperDbContext.Create);
-            app.CreatePerOwinContext<AppUserManager>(AppUserManager.Create);
-            app.CreatePerOwinContext<AppSignInManager>(AppSignInManager.Create);
-
             // Configure the sign in cookie
             app.UseCookieAuthentication(new CookieAuthenticationOptions
```
Comment (goremukin) [inline PortfolioKeeper/App_Start/Startup.Auth.cs]: Move 'new CookieAuthenticationOptions' to the new line, please

File: PortfolioKeeper/App_Start/Startup.Auth.cs [hunk 3]
```diff
@@ -14,10 +14,6 @@ public partial class Startup
...
             {
```
Comment (goremukin) [inline PortfolioKeeper/App_Start/Startup.Auth.cs]: Add tab to 23-35 lines, please

File: PortfolioKeeper/BusinessLogic/DtoModels/PortfolioDto.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace PortfolioKeeper.BusinessLogic.DtoModels
+{
+    public class PortfolioDto
+    {
+        public int Id { get; set; }
```
Comment (goremukin) [inline PortfolioKeeper/BusinessLogic/DtoModels/PortfolioDto.cs]: Add here and below the empty line, please

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+{
+    public class PortfolioService : IPortfolioService
+    {
+        private readonly IUnitOfWork unitOfWork;
+        private readonly IMapper mapper;
+        private readonly IRepository<Portfolio, int> portfolioRepository;
+
+        public PortfolioService(IUnitOfWork unitOfWork, IMapper mapper)
+        {
+            if (unitOfWork == null)
```
Comment (goremukin) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: Please use this. It looks much more compact
`this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));`
`this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));`

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            }
+
+            this.unitOfWork = unitOfWork;
+            this.mapper = mapper;
+            portfolioRepository = unitOfWork.GetRepository<Portfolio, int>();
+        }
+
+        public IEnumerable<string> Create(PortfolioDto portfolioDto)
+        {
+            if (portfolioDto == null)
```
Comment (goremukin) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: and here

File: PortfolioKeeper/Controllers/AccountController.cs [hunk 7]
```diff
@@ -14,24 +17,16 @@ namespace PortfolioKeeper.Controllers
...
-        {
-        }
+        public AppSignInManager SignInManager => signInManager;
+
+        public AppUserManager UserManager => userManager;
 
-        public AccountController(AppUserManager userManager, AppSignInManager signInManager)
+        public AccountController(AppUserManager userManager, AppSignInManager signInManager, IPortfolioService portfolioService)
         {
             if (userManager == null)
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/AccountController.cs]: and here

File: PortfolioKeeper/Controllers/AccountController.cs [hunk 8]
```diff
@@ -41,8 +36,13 @@ public AccountController(AppUserManager userManager, AppSignInManager signInMana
             {
                 throw new ArgumentNullException(nameof(signInManager));
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/AccountController.cs]: and here

File: PortfolioKeeper/Controllers/AccountController.cs [hunk 9]
```diff
@@ -41,8 +36,13 @@ public AccountController(AppUserManager userManager, AppSignInManager signInMana
...
             }
-            UserManager = userManager;
-            SignInManager = signInManager;
+            if (portfolioService == null)
+            {
+                throw new ArgumentNullException();
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/AccountController.cs]: and here

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 10]
```diff
@@ -15,25 +14,25 @@ public class ManageController : Controller
...
-        {
-            get => signInManager ?? HttpContext.GetOwinContext().Get<AppSignInManager>();
-            private set => signInManager = value;
-        }
-        public AppUserManager UserManager
-        {
-            get => userManager ?? HttpContext.GetOwinContext().GetUserManager<AppUserManager>();
-            private set => userManager = value;
-        }
+        public AppSignInManager SignInManager => signInManager;
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/ManageController.cs]: Add empty line, please

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 11]
```diff
@@ -15,25 +14,25 @@ public class ManageController : Controller
...
         {
         }
 
         public ManageController(AppUserManager userManager, AppSignInManager signInManager)
         {
-            UserManager = userManager;
-            SignInManager = signInManager;
+            if (userManager == null)
+            {
+                throw new ArgumentNullException(nameof(userManager));
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/ManageController.cs]: Use a more compact format, please

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 12]
```diff
@@ -15,25 +14,25 @@ public class ManageController : Controller
...
+            }
+            if (signInManager == null)
+            {
+                throw new ArgumentNullException(nameof(signInManager));
```
Comment (goremukin) [inline PortfolioKeeper/Controllers/ManageController.cs]: and here

File: PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs [hunk 13]
```diff
@@ -9,7 +9,7 @@ public class PortfolioKeeperDbContext : IdentityDbContext
     {
         static PortfolioKeeperDbContext()
         {
-            Database.SetInitializer(new MigrateDatabaseToLatestVersion<PortfolioKeeperDbContext, Migrations.Configuration>());
+           // Database.SetInitializer(new MigrateDatabaseToLatestVersion<PortfolioKeeperDbContext, Migrations.Configuration>());
```
Comment (goremukin) [inline PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs]: You can delete it

File: PortfolioKeeper/App_Start/Startup.Auth.cs (hunk 2 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/App_Start/Startup.Auth.cs]: fixed

File: PortfolioKeeper/App_Start/Startup.Auth.cs (hunk 3 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/App_Start/Startup.Auth.cs]: done

File: PortfolioKeeper/BusinessLogic/DtoModels/PortfolioDto.cs (hunk 4 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/BusinessLogic/DtoModels/PortfolioDto.cs]: done

File: PortfolioKeeper/Controllers/AccountController.cs (hunk 8 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/Controllers/AccountController.cs]: done

File: PortfolioKeeper/Controllers/AccountController.cs (hunk 9 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/Controllers/AccountController.cs]: done

File: PortfolioKeeper/Controllers/ManageController.cs (hunk 10 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/Controllers/ManageController.cs]: fixed

File: PortfolioKeeper/Controllers/ManageController.cs (hunk 11 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/Controllers/ManageController.cs]: done

File: PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs (hunk 13 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/Data/Context/PortfolioKeeperDbContext.cs]: Master will overwrite this file

File: PortfolioKeeper/App_Start/AutofacConfig.cs (hunk 1 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/App_Start/AutofacConfig.cs]: Сhanged

File: PortfolioKeeper/App_Start/AutofacConfig.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+        public static IContainer CreateContainer(IAppBuilder app)
+        {
+            var builder = new ContainerBuilder();
+            RegisterDalComponents(builder);
+            RegisterBusinessComponents(builder);
+            RegisterUIComponents(builder);
+            RegisterHelperComponents(builder);
+
+            #region identityComponents
```
Comment (EdwOK) [inline PortfolioKeeper/App_Start/AutofacConfig.cs]: move to `Identity` method

File: PortfolioKeeper/Controllers/AccountController.cs [hunk 15]
```diff
@@ -222,17 +226,8 @@ protected override void Dispose(bool disposing)
...
-                    userManager.Dispose();
-                    userManager = null;
-                }
-
-                if (signInManager != null)
-                {
-                    signInManager.Dispose();
-                    signInManager = null;
-                }
+                userManager.Dispose();
```
Comment (EdwOK) [inline PortfolioKeeper/Controllers/AccountController.cs]: null propagation operator `?.`

File: PortfolioKeeper/Controllers/ManageController.cs [hunk 16]
```diff
@@ -155,10 +144,9 @@ public ActionResult SetPassword()
 
         protected override void Dispose(bool disposing)
         {
-            if (disposing && userManager != null)
+            if (disposing)
             {
                 userManager.Dispose();
```
Comment (EdwOK) [inline PortfolioKeeper/Controllers/ManageController.cs]: null propagation operator `?.`

File: PortfolioKeeper/Models/AccountViewModels.cs [hunk 17]
```diff
@@ -66,7 +76,7 @@ public class ResetPasswordViewModel
 
         [DataType(DataType.Password)]
         [Display(Name = "Confirm password")]
-        [Compare("Password", ErrorMessage = "The password and confirmation password do not match.")]
+        [System.ComponentModel.DataAnnotations.Compare("Password", ErrorMessage = "The password and confirmation password do not match.")]
```
Comment (EdwOK) [inline PortfolioKeeper/Models/AccountViewModels.cs]: `namespace System.ComponentModel.DataAnnotations`

**Comment by AntonBessarabov** [comment]
All fixed

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+            }
+            var result = new List<string>();
+            try
+            {
+                var newPlot = mapper.Map<PortfolioDto, Portfolio>(portfolioDto);
+                newPlot.CreatedOn = newPlot.UpdatedOn = DateTime.UtcNow;
+                portfolioRepository.Create(newPlot);
+                unitOfWork.Commit();
+            }
+            catch (Exception e)
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: this is a bad approach, specify which errors may be thrown

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        private readonly IUnitOfWork unitOfWork;
+        private readonly IMapper mapper;
+        private readonly IRepository<Portfolio, int> portfolioRepository;
+
+        public PortfolioService(IUnitOfWork unitOfWork, IMapper mapper)
+        {
+            this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
+            this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+            portfolioRepository = unitOfWork.GetRepository<Portfolio, int>() ?? throw new ArgumentNullException(nameof(portfolioRepository));
+
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: extra space

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public IEnumerable<string> Create(PortfolioDto portfolioDto)
+        {
+            if (portfolioDto == null)
+            {
+                throw new ArgumentNullException(nameof(portfolioDto));
+            }
+            var result = new List<string>();
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: space around

File: PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs (hunk 18 above)
Comment (AntonBessarabov) [reply PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: unitOfWork.Commit() throws 6 exceptions:

-DbUpdateException : DataException -> DataException : SystemException
-DbUpdateConcurrencyException : DbUpdateException
-DbEntityValidationException : DataException

-NotSupportedException : SystemException -> SystemException : Exception
-ObjectDisposedException : InvalidOperationException
-InvalidOperationException : SystemException

What exceptions do you propose to intercept?

Comment (AntonBessarabov) [reply PortfolioKeeper/BusinessLogic/Services/PortfolioService.cs]: Or just handle SystemExcpetion-s here?

**Pull request merged**